dashboard/app/models/course.rb

```diff
@@ -121,4 +121,57 @@ def summarize
       end
     }
   end
+
+  @@course_cache = nil
+  COURSE_CACHE_KEY = 'course-cache'.freeze
+
+  def self.should_cache?
+    return false if Rails.application.config.levelbuilder_mode
+    ENV['UNIT_TEST'] || ENV['CI']
+  end
+
+  # generates our course_cache from what is in the Rails cache
+  def self.course_cache_from_cache
+    Course.connection
+    # make sure possible loaded objects are completely loaded
+    [CourseScript, Plc::Course].each(&:new)
+    Rails.cache.read COURSE_CACHE_KEY
+  end
+
+  def self.course_cache_from_db
+    {}.tap do |cache|
+      Course.all.pluck(:id).each do |course_id|
+        course = get_without_cache(course_id)
+        cache[course.name] = course
+        cache[course.id.to_s] = course
+      end
+    end
+  end
+
+  def self.course_cache_to_cache
+    Rails.cache.write(COURSE_CACHE_KEY, course_cache_from_db)
+  end
+
+  def self.course_cache
+    return nil unless should_cache?
+    @@course_cache ||=
+      course_cache_from_cache || course_cache_from_db
+  end
+
+  def self.get_without_cache(id_or_name)
+    # a bit of trickery so we support both ids which are numbers and
+    # names which are strings that may contain numbers (eg. 2-3)
+    find_by = (id_or_name.to_i.to_s == id_or_name.to_s) ? :id : :name
```

Comment:
I'm ok leaving this alone now, and revisiting when (if) we clean this up and consolidate the caching logic. I can help with that effort. 

FWIW, we use this same regex for a similar purpose in other [places](https://github.com/code-dot-org/code-dot-org/blob/513edbf4e9855e00d0e9d77c2e135a0514ff2578/dashboard/app/controllers/pd/teacher_application_controller.rb#L31).